packages/context/src/resolver.ts

```diff
@@ -136,6 +136,77 @@ function resolve<T>(
   return resolved;
 }
 
+/**
+ * Resolve entries of an object into a new object with the same keys or a
+ * promise of the new object. Please note some of the entries can be resolved
+ * synchronously while others need to be resolved asynchronously.
+ * @param map The original object containing the source entries
+ * @param resolver A function resolves an entry to a value or promise
+ */
+export function resolveMap<T, V>(
+  map: {[name: string]: T},
+  resolver: (v: T, map: {[name: string]: V}, p: string) => ValueOrPromise<V>,
+) {
+  const result: {[name: string]: V} = {};
+  let asyncResolvers: PromiseLike<void>[] | undefined = undefined;
+
+  const setter = (p: string) => (v: V) => {
+    result[p] = v;
+  };
+
+  for (const p in map) {
+    const valueOrPromise = resolver(map[p], result, p);
+    if (isPromise(valueOrPromise)) {
+      if (!asyncResolvers) asyncResolvers = [];
+      asyncResolvers.push(valueOrPromise.then(setter(p)));
+    } else {
+      result[p] = valueOrPromise;
+    }
+  }
+
+  if (asyncResolvers) {
+    return Promise.all(asyncResolvers).then(() => result);
+  } else {
+    return result;
+  }
+}
+
+/**
+ * Resolve entries of an array into a new array with the same indexes or a
+ * promise of the new object. Please note some of the entries can be resolved
+ * synchronously while others need to be resolved asynchronously.
+ * @param list The original array containing the source entries
+ * @param resolver A function resolves an entry to a value or promise
+ */
+export function resolveList<T, V>(
+  list: T[],
+  resolver: (v: T, list: V[], i: number) => ValueOrPromise<V>,
+) {
```

Comment:
Could you please add an explicit type signature of the returned value? I know TypeScript is able to infer the type for us, but I think the readers of this code will appreciate the explicit type.